keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,418 @@
+- Redesign APIs/refactor the existing set of parameters. This is about providing
+  a clear migration path for all classes of parameters.
+- Eliminate instance-specific config parameters altogether. This is a special
+  case of the previous item.
+- Make changes to the Dynamic kubelet Config feature or address difficulties in
+  using the kubelet's current ComponentConfig API with Dynamic kubelet Config.
+
+## Proposal
+
+This KEP has two major parts. First, a simple solution to cleanly separate
+instance-specific parameters from shareable parameters so that the latter may
+continue to be shared via a single ConfigMap. Second, provide examples of how
+to configure instance-specific parameters in the new model for kubelet and
+kube-proxy, which are the two components currently capable of sharing parameters
+from the same ConfigMap.
+
+### Separate config object for instance-specific config
+
+The proposal is this: Add a new Kind for instance-specific config, and put
+instance-specific parameters in that object.
+
+This is relatively simple and isolates the parameters so that they can be
+dealt with separately. ComponentConfig files containing only sharable parameters
+can continue to be shared via a single ConfigMap. Files containing only
+instance-specific parameters can be provided to components via other means,
+such as the node startup scrips or an init container that inserts values from
+the Pod's Downward API.
+
+### Why not just keep using flags for instance-specific parameters?
+
+Flags have various problems, outlined in the [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c)
+doc.
+
+Leaving some parameters as flags while the rest are exposed via ComponentConfig
+results in an inconsistent API surface, and an inconsistent API versioning
+policy. Why fix only part of the problem when we can fix the whole thing?
+
+### Why not just auto-detect?
+
+One alternative that has been proposed in the past is to find a way to
+auto-detect values of instance-specific parameters and simply eliminate them
+from the configuration workflow. This is a great idea in theory, because it
+solves the problem by removing work for users, but it may be difficult in
+practice. For example, on a machine with multiple network cards, which IP should
+the kubelet use?
+
+Implementing the two-object approach does not prevent us from finding ways to
+auto-detect instance specific config in the future. It is worth noting that the
+_possibility_ of someday being able to auto-detect these has led to some
+hesitation to make a firm decision today. We have delayed for too long already,
+and a firm decision is needed to move forward.
+
+### Risks and Mitigations
+
+This creates an additional bucket for config parameters. We must be vigilant
+that only parameters that are truly instance-specific are added to the
+instance-specific Kind. We can provide warnings to developers via comments
```

Comment:
missing in the proposal above: the new kind defines which parameters are instance-specific and which are not.

Also: is there an overlap?

Reply:
Please at that very centrally in the summary and the proposed solution. IMO this is the key idea of the proposal.